Add the `interestfor` attribute

foolip opened the pull request:
The `interestfor` attribute indicates that an element is an interest
invoker and points (via ID) at the element that should be invoked (made
visible, typically) when the user shows interest in the invoker. When
the target is a popover, it is automatically opened and closed as
interest is gained and lost. The delay before interest is gained or lost
can be customized using the `interest-delay-*` CSS properties.

The typical use case for this is hovercards and tooltips.

Keyboards, pointing devices, and touch behavior is normatively defined, and the
user agent should provide a way for the user to express interest regardless of
input modality.

Based on https://open-ui.org/components/interest-invokers.explainer/ and
the implementation in Chromium.

The `:interest-source` and `:interest-target` CSS pseudo-classes are defined here:
https://drafts.csswg.org/selectors/#interest-pseudos

The `interest-delay-*` CSS properties are defined here:
https://drafts.csswg.org/css-ui-4/#interest

***

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.

When you submit this PR, and each time you edit this comment (including checking a checkbox through the UI!), PR Preview will run and update it. As such make any edits in one go and only after PR Preview has run.

If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [ ] At least two implementers are interested (and none opposed):
   * Chromium ([Intent to Prototype](https://groups.google.com/a/chromium.org/g/blink-dev/c/UO30brSlWhs/m/gfULRkmYAgAJ), [Intent to Experiment](https://groups.google.com/a/chromium.org/g/blink-dev/c/LLgsMjTzmAY/m/o6ymyzauAQAJ), [Intent to Extend Experiment](https://groups.google.com/a/chromium.org/g/blink-dev/c/0aqsckMRjH8/m/PQ4MfNugDQAJ)) 
   * …
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/semantics/the-button-element/interest-for
   * https://chromium-review.googlesource.com/c/chromium/src/+/6908595
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://crbug.com/326681249
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1984699
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=297786
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): N/A
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): N/A
- [x] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs: https://github.com/w3c/aria/pull/2617
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/mdn/issues/721
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/11006/browsers.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/browsers.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/browsers.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/form-elements.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/form-elements.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/image-maps.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/image-maps.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/image-maps.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/index.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/index.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/index.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/indices.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/indices.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/indices.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/infrastructure.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/infrastructure.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/interaction.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/interaction.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/interaction.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/popover.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/popover.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/popover.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/semantics-other.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/semantics-other.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/semantics-other.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/text-level-semantics.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/text-level-semantics.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/text-level-semantics.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )
<a href="https://whatpr.org/html/11006/interestfor.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">/interestfor.html</a>  ( <a href="https://whatpr.org/html/11006/a1c1bd1...fa121b6/interestfor.html" title="Last updated on Oct 1, 2025, 11:59 AM UTC (fa121b6)">diff</a> )

foolip committed:
Add the interesttarget attribute

foolip commented:
This is an initial translation of https://open-ui.org/components/interest-invokers.explainer/ to spec language, with a lot of things incomplete.

@mfreed7 I think the two biggest points of uncertainty I have now are the integration with mouse/keyboard/touch input on the one end, and with CSS on the other.

For showing interest, the timing of these algorithms relative to "mouseover" and other events needs to be defined.

For CSS, the fact that there are CSS properties controlling the timing means we need to think about:

- When do we look at the computed style to extract those times?
- Is there a state machine with timers, and is anything observable between the initial sign of interest and when we run "capture interest"?

There are many more minor TODOs, but I'd like to get the high-level flow settled first.

mfreed7 commented on source:
```diff
@@ -74941,6 +74947,13 @@ contradict people?
     data-x="concept-selector-active">being activated</i>.)</p>
    </dd>
 
+   <dt><dfn selector noexport><code data-x="selector-has-interest">:has-interest</code></dfn></dt>
```
Does this belong here in HTML? Or better only in CSS?

foolip replied:
The typical setup is that https://drafts.csswg.org/selectors/ defines that the selectors exist, and then HTML defines when it matches for HTML. The idea is that HTML is a "host language" and there can be others, like SVG.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <h4>The <code data-x="attr-interesttarget">interesttarget</code> attribute</h4>
+
+  <p>The <dfn element-attr for="html-global"><code data-x="attr-interesttarget">interesttarget</code></dfn> attribute on <code>a</code>, <code>area</code>, and <code>button</code> elements TODO.</p>
```
I think the TODO here is to describe what they do? If so, let me know if you want help writing that.

foolip replied:
That's right, a non-normative description should go here. I would have copied https://html.spec.whatwg.org/multipage/popover.html#the-popover-target-attributes, but there is no description to introduce the attributes there.

https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute is a good example I think.

Happy to take your wording if you have one in mind!

mfreed7 replied:
Interesting. I went looking for the `command` and `commandfor` attributes version of this, and there isn't even a section like this. Seems inconsistent.

Here's a quick stab at this one for interesttarget:

```
The interesttarget attribute on a, area, and button elements allows authors to set up an invoker relationship between the triggering element and a separate target element such as a popover. With this arrangement, when the user shows interest in the triggering element (e.g. by hovering or focusing it), the target element will have an `interest` event fired on it. If the target is a popover, this will show the popover.
```

foolip replied:
Thanks, I've pushed that description.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+&lt;/button>
+
+&lt;div popover=hint id=tooltip>
+ I will appear
```
nit: "I will appear when the user shows interest in the button" or something?

foolip replied:
Let's go with that!

```suggestion
 I will appear when the user shows interest in the button
```

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+&lt;/div></code></pre>
+ </div>
+
+ <p>Every <span data-x="HTML elements">HTML element</span> has a <dfn>has interest</dfn>, which is a
```
I know this is the HTML spec, but the plan at least was to minimally support `<svg>` `<a>` elements. So perhaps this should be just `Element`?

foolip replied:
Popovers were only made to work for HTML elements, if I'm reading the spec correctly. Would just `interesttarget` without fully working popovers be useful, or would it seem broken? (We can fix both, if it would be useful.)

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+ <p>Every <span data-x="HTML elements">HTML element</span> has a <dfn>has interest</dfn>, which is a
+ boolean, initially set to false.</p>
+ <!-- TODO: or have an Element slot on Document for this since there can only be one at a time? -->
```
Nope - there can be more than one at a time. The `interest-target-hide-delay` being non-zero and less than the `-show-delay` value means more than one element can have interest at a time.

foolip replied:
I see, I'll delete this TODO.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  data-x="HTML elements">HTML element</span> <var>invoker</var>:</p>
+
+  <ol>
+   <li><p>Assert: <var>invoker</var> is an <code>a</code>, <code>area</code>, or <code>button</code>
```
...or SVG <a>

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>Let <var>continue</var> be the result of <span data-x="concept-event-fire">firing an
+   event</span> named <code data-x="event-interest">interest</code> at <var>target</var>, using
+   <code>InterestEvent</code>, with the <code data-x="dom-Event-cancelable">cancelable</code> and
+   <code data-x="dom-Event-composed">composed</code> attributes initialized to true, and the
```
Per recent discussions about `CommandEvent`, perhaps this should be `composed=false`.

foolip replied:
Done.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <code data-x="dom-Event-composed">composed</code> attributes initialized to true, and the
+   <code data-x="dom-InterestEvent-source">source</code> attribute initialized to
+   <var>invoker</var>.</p></li>
+   <!-- TODO: the explainer calls for composed without bubbling, does that make sense? -->
```
I just fixed the explainer to say `non-composed`.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>If <var>target</var>'s <code data-x="attr-popover">popover</code> attribute is not in the
+   <span data-x="attr-popover-none-state">no popover state</span>, then run <span>show popover</span>
+   given <var>target</var>, false, and <var>invoker</var>.</p></li>
+   <!-- TODO: should invoker be passed? It's used for popovertarget, might not do the right thing
```
Yes, it should definitely be passed here. It'll do the right thing. 😄

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <!-- TODO: should invoker be passed? It's used for popovertarget, might not do the right thing
+              if passed in here... -->
+
+   <!-- TODO: nothing to do if not a popover? how can devs show/hide it? needs another pseudo-class? -->
```
Other than fire the event, nothing else to do.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+    false.</p></li>
+
+    <li><p>Set <var>invoker</var>'s <span>has interest</span> to false.</p></li>
+    <!-- TODO: is this the right time to change the state? too late? -->
```
I think this should likely go above the `hide popover` call, since the event handlers there should be able to see that the interest has been lost.

mfreed7 commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <ul>
+   <li><p>Actually invoke the capture/lose algorithms.</p></li>
+   <li><p>Which algorithm reads the computed style for interest-target-delay?</p></li>
```
There need to be algorithms equivalent to these from Chromium code:

- [HandleInterestTargetHoverOrFocus](https://source.chromium.org/search?q=Element::HandleInterestTargetHoverOrFocus&sq=&ss=chromium%2Fchromium%2Fsrc)
- [ScheduleInterestGainedTask](https://source.chromium.org/search?q=Element::ScheduleInterestGainedTask&ss=chromium%2Fchromium%2Fsrc)
- [ScheduleInterestLostTask](https://source.chromium.org/search?q=Element::ScheduleInterestLostTask&sq=&ss=chromium%2Fchromium%2Fsrc)

I think those three should hit most of the things in this list. Hopefully.

foolip commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+ </div>
+
+ <p>Every <span data-x="HTML elements">HTML element</span> has a <dfn>has interest</dfn>, which is a
+ boolean, initially set to false.</p>
```
Ah, a bunch of changes in https://github.com/openui/open-ui/pull/1213, I'll look through them.

foolip replied:
Do you have a preferred name for this internal state? The one that comes to mind is "interest state". And should it be exposed as an IDL attribute and not just through the matching of pseudo-classes?

foolip commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  <pre><code class="idl">[Exposed=Window]
+interface <dfn interface>InterestEvent</dfn> : <span>Event</span> {
+  constructor(DOMString type, optional <span>InterestEventInit</span> eventInitDict = {});
+  readonly attribute Element? source; // TODO: nullable, or require in ctor?
```
I agree, that's more common in recent event interfaces, I'll make the change. (Otherwise, it's nullable just because of the constructor, which is rarely used anyway.)

foolip commented on source:
```diff
@@ -87118,6 +87131,159 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+ <div class="example">
+  <p>The following demonstrates how one might show a tooltip for a button using the
+  <code data-x="attr-interesttarget">interesttarget</code> attribute to associate the button with
+  a <code>div</code> <code data-x="attr-popover">popover</code> representing the tooltip.</p>
```
Fair enough. "div popover" did make it sound a bit like "popover" was a variable and was a `<div>`.

foolip committed:
Address the easy review feedback

foolip committed:
Add additional pseudo-classes

foolip committed:
make InterestEvent non-composed

foolip committed:
add interesttarget attr description

foolip committed:
Make interest tri-state none/partial/full

foolip committed:
Fix TODOs

foolip committed:
Define :target-of-interest / :target-of-partial-interest

foolip commented on source:
```diff
@@ -88115,6 +88156,166 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+ <ul>
+  <li><p><dfn export for="interest state" data-x="interest-state-none">none</dfn></p></li>
+
+  <li><p><dfn export for="interest state" data-x="interest-state-partial">partial</dfn></p></li>
```
@mfreed7 Chromium additionally has a "potential partial interest" state used for popovers:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=11060-11067;drc=743a82d08e59d803c94ee1b8564b8b11dd7b462f

Do we have to do the same in the spec? Or can we check if the target element is a popover when the scheduled task runs? The `popover` attribute could be removed before the task runs, so checking it ahead of time seems like it adds the need to double-check it later.

mfreed7 replied:
Per other conversations, this whole thread is moot.

mfreed7 reviewed (commented):
Ok, I went through these fairly carefully and tried to provide links to Chromium code for the missing stuff. Hopefully that makes it a lot easier for you to find the equivalent code that we're trying to spec out.

foolip commented on source:
```diff
@@ -75929,6 +75950,26 @@ Demos:
 
+   <dt><dfn selector noexport><code data-x="selector-target-of-interest">:target-of-interest</code></dfn></dt>
+   <dd>
+    <p>The <code data-x="selector-target-of-interest">:target-of-interest</code>
```
The way this is written here clearly isn't how you can implement it. I assume the Chromium implementation keeps tracks of the extra state to implement the pseudo-class efficiently?

foolip commented on source:
```diff
@@ -144434,6 +144635,13 @@ interface <dfn interface>External</dfn> {
+          <code data-x="attr-interesttarget">area</code>;
+          <code data-x="attr-interesttarget">button</code>
+     <td> Targets an interesting TODO
+     <td> <span data-x="concept-id">ID</span>* <!-- TODO: is the asterisk needed, is it really complicated? -->
```
I noticed that some things in https://html.spec.whatwg.org/multipage/indices.html#attributes-3 have an asterisk, and below the table it says:

> An asterisk (*) in a cell indicates that the actual rules are more complicated than indicated in the table above.

I'm not sure why it's ID* everywhere instead of ID, I think I'll have to ask another HTML spec editor. Or just leave the asterisk in.

mfreed7 commented on source:
```diff
@@ -88115,6 +88156,166 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <ul>
+   <li><p>Actually invoke the capture/lose algorithms.</p></li>
+   <li><p>Which algorithm reads the computed style for interest-target-delay?</p></li>
```
Latest names:
- `interest-delay` (shorthand)
- `interest-show-delay`
- `interest-hide-delay`

https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2940736593

foolip committed:
Apply Mason's review suggestions

foolip committed:
Rename to interestfor

foolip committed:
Remove pseudo-classes

foolip commented on source:
```diff
@@ -88115,6 +88156,166 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   given <var>target</var>, false, and <var>invoker</var>.</p></li>
+  </ol>
+
+  <p>To <dfn export data-x="lose-interest">lose interest</dfn>, given an <span
```
Sure, I'll make that change for both.

foolip committed:
Apply additional review feedback

mfreed7 commented on source:
```diff
@@ -88115,6 +88156,166 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>Set <var>invoker</var>'s <span>interest state</span> to
+   <span data-x="interest-state-full">full</span>.</p></li>
+
+   <!-- TODO: test that :has-interest doesn't match when the interest event is fired, but does
```
I believe that is hand-wavingly tested, but I'll take an AI to write a specific test for the timing of `interest` and `:has-interest`.

mfreed7 replied:
https://chromium-review.googlesource.com/c/chromium/src/+/6697154

foolip committed:
Start wiring up focus/blur events

foolip committed:
Fix typo

foolip committed:
Add error handling

foolip committed:
Use likely new state as argument

foolip committed:
Remove TODO (https://chromium-review.googlesource.com/c/chromium/src/+/6697154)

foolip commented on source:
```diff
@@ -88115,6 +88156,166 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+    <li><p>If <var>target</var> is null, then return.</p></li>
+
+    <!-- TODO: what state is possibly left dangling if the target wasn't found? -->
```
I did some testing of when the `interestfor` link breaks and filed https://github.com/openui/open-ui/issues/1240 with a demo.

mfreed7 replied:
Yep, we discussed at OpenUI and decided that it's just best (and most parallel to other things like `popover`) to just break the links and update CSS pseudo classes, but don't fire any events or anything.

foolip committed:
Add beginnings of attribute/tree change handling

foolip committed:
Wire up mouse interaction

foolip committed:
Fix a TODO

foolip committed:
Remove the TODO for ID*

The extra requirement is:

> must be the ID of an element in the same tree as the element with the interestfor attribute

foolip committed:
Fix mouse event names

foolip committed:
Fix TODOs by adding active interest target/source

foolip committed:
Avoid stale active interest source

foolip committed:
Introduce pending interest change handle to cancel tasks

foolip committed:
Merge remote-tracking branch 'origin/main' into foolip/interesttarget

foolip committed:
Add missing target arguments

foolip committed:
Drop partial interest state for now

foolip committed:
Add a should-level requirement for other modalities

mfreed7 commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>If <var>namespace</var> is not null, then return.</p></li>
+
+   <li><p>If <var>localName</var> is not <code data-x="attr-id">id</code>, then return.</p></li>
+
```
Does there need to be a step that makes sure the `id` value has actually changed? (We do that in Chromium, but perhaps the spec only calls this algo when the value actually changes. So just checking.)

foolip replied:
I've also added the same check for `interestfor` but it looks like Chromium doesn't have that check:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=3464-3473;drc=e18a32c4b96ad2b2c2f50195936594d1f7c4f62a

Setting `id` and `interestfor` to the same value both need testing in WPT.

mfreed7 commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li>
+    <p><span>Reset interest state</span> for <var>invoker</var>.</p>
+
+    <p class="note">This ensures that if <var>invoker</var>'s <span>active interest target</span>
```
So Chromium code has a bit  more logic to handle this case, where something has an existing invoker, here:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=11113;drc=59941535ecc1a62e988608ad8bab0dbdbb730eb5

I think it might be good to include that logic, since it explicitly handles firing `loseinterest` at the old target, and some other cleanups. And it happens before firing the new `interest` event also, which I also think is important, so that JS code in that event sees the correct current state of things. And this can then be converted to an `assert` that invoker doesn't have interest.

foolip replied:
The 3 events are the 2 "loseinterst" events and then the "interest" event. Checking after each seems sensible to me, in which case I should define a helper for those checks.

mfreed7 replied:
Yeah, we should likely check after each one, for safety?

mfreed7 commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <p class="note">Keyboard interactions are handled in the <span>focus update steps</span>.</p>
+
+  <p>For input modalities other than pointing devices and keyboards, the user agent should provide
```
It might be worth explicitly saying that for touchscreen interfaces that support the long-press gesture:

1. If a long-press would otherwise show a context menu, add an item to the context menu that allows the user to show interest in the element.
2. If a long-press wouldn't show any context menu, then long-pressing the element shows interest immediately (skipping delays).

foolip replied:
Yeah, I think it would be best if canceling the "mouseover" event has no effect on this feature. https://software.hixie.ch/utilities/js/live-dom-viewer/?saved=13944 shows that it doesn't prevent `:hover` from matching, and I've defined the trigger for interest in the spec using the same terms as `:hover`.

A developer that wants to can always cancel the "interest" event still.

mfreed7 commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <p>The <dfn attribute for="InterestEvent"><code
+  data-x="dom-InterestEvent-source">source</code></dfn> attribute must return the value it was
+  initialized to.</p>
```
Is this the typical way this is worded? It feels a little weird. Like either it doesn't need to be said, or there's another way? (Maybe it's fine, it just sounds odd.)

foolip replied:
Yeah, it's a bit antiquated, but still common for events. The "better" way would be to define slots for everything and let the getters use those slots, but it's a lot of boilerplate.

@annevk I'm not sure if there is a more modern pattern I could copy, I'd be happy to as long as I'm not the pioneer :)

mfreed7 commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  the triggering element (e.g., by hovering or focusing it), the target element will have an <code
+  data-x="event-interest">interest</code> event fired on it. If the target is a popover with a
+  <span>popover visibility state</span> of <span data-x="popover-hidden-state">hidden</span>, this
+  will show the popover.</p>
```
Do we want to add another sentence that says that the popover will be hidden when the user loses interest? And mention the `loseinterest` event?

mfreed7 reviewed (commented):
Some comments, but overall I think this looks great!

mfreed7 commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  <var>element</var> with a pointing device, the user agent must <span>queue a task</span> on the
+  <span>user interaction task source</span> to <span data-x="handle-interest-change">handle interest
+  change</span> for <var>element</var> and <span data-x="interest-state-none">none</span>.</p>
+
```
I realized that I think there also needs to be a call to `handle interest change` from wherever we handle the default behavior for a keypress, for the `ESC` key. This is done in Chromium [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=1745;drc=c780b245b236c9240d71797169acc36e3b6e1c63).

On a similar topic, there's also [code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_element.cc;l=1908;drc=59941535ecc1a62e988608ad8bab0dbdbb730eb5) in Chromium that immediately loses interest when the target popover is closed via other means. See the [CL description](https://chromium-review.googlesource.com/c/chromium/src/+/6536349) for more detail, but this is important to avoid out-of-sync UX problems, which are pretty easy to hit without this behavior.

mfreed7 replied:
> If I'm reading and testing correctly, hitting Esc should lose interest without a delay and should fire the "loseinterest" event. But it looks like the event is cancelable but cannot actually be canceled. What do you think we should do here, fire an event that isn't cancelable? Also needs tests.

Oh interesting! I hadn't handled that either. I do think the `loseinterest` event that gets fired as a result of the user hitting ESC shouldn't be cancelable. ESC is really the escape hatch and should always work in my opinion.

> In the spec, I think a close watcher (like [popover close watcher](https://html.spec.whatwg.org/multipage/popover.html#popover-close-watcher)) will be needed here, definitely for the case where the target isn't a popover.

Does this need a close watcher? Since this part is about interest, which may or may not trigger a popover, I'm not sure we need all of that machinery. The popover already has a closewatcher and that should handle most popover use cases. Other than that I just think ESC should lose interest, but perhaps e.g. android back button should not? I.e. I'd expect the back button to go immediately back, not first lose interest, and then second press goes back. That's different than a popover or dialog.

> But when the target is a popover, should the "toggle" event or the "loseinterest" event come first? This will also need testing.

I'd still expect the `loseinterest` to come first. Just that it's not cancelable, so there's always a `beforetoggle` and `toggle` after that, assuming the `loseinterest` handler doesn't remove the popover or something.

mfreed7 replied:
The issue I'm worrying about is that closewatcher comes with a lot of developer-visible stuff, like the ability to add your own closewatcher that interacts with the close watcher stack, etc. It also makes the e.g. Android Back button trigger it. I'm not sure we want/need that stuff here. Like I think Android Back button should just go directly back, and not first trigger a loss of interest.

foolip replied:
I have added something in the close request steps now.

The main question I had was whether _closedSomething_ should be set to true due to losing interest and thus prevent the "alternative processing" steps. Under the assumption that Android Back falls under "alternative processing" I decided not to. This means that we'll both (try to) lose interest and navigate away due to a single <kbd>Esc</kbd> press. With bfcache this means that interest will be lost (unless the "loseinterest" event is canceled) when navigating back, so that things like hovercards aren't showing if navigating forward again. I think this makes sense, but want to confirm.

@mfreed7 can you look at the change (commit 5d49fa2787d4988c18283d9117123b814e664dd7) and see if you think it's the right behavior?

mfreed7 replied:
Yeah I see what you mean - that's not how it'll be implemented. But it makes sense and is clear, so I think it should work as far as spec text goes.

foolip replied:
Alright, let's try this as a first approach then! This was the last remaining bit, so now I'll just wait for review from more folks.

foolip committed:
Merge remote-tracking branch 'origin/main' into foolip/interesttarget

foolip committed:
Fix the moving steps

foolip committed:
Use active interest pointers in removal and attribute change steps

foolip committed:
Simplify away the interest state slot/enum

foolip commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+ <ul>
+  <li><p><dfn export for="interest state" data-x="interest-state-none">none</dfn></p></li>
+
+  <li><p><dfn export for="interest state" data-x="interest-state-full">full</dfn></p></li>
```
We can go even further and just use _active interest target_ as the state. It's not needed as a slot on HTML element, and the remaining use as an argument for "handle interest change" didn't feel like enough to keep around an enum. A _show_ boolean aligned pretty well with the switch to pick which CSS property to use.

mfreed7 replied:
Makes sense to me!

foolip committed:
Remove unnecessary data-x and export

foolip committed:
Handle interest change for targets, not just invokers

foolip commented on source:
```diff
@@ -87993,6 +88032,309 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   the attr-associated element">get the <code data-x="">interestfor</code>-associated
+   element</span>.</p></li>
+
+   <li><p>If <var>target</var> is null, then return.</p></li>
```
Ah, of course! I've spec'd this as a recursive call to "handle interest change" so that it should have exactly the same effect. Is there any nuance here, anything that shouldn't be done? (There are bits that can be optimized since we already know the target and don't need to look it up using `interestfor`, but that can be an exercise for the reader.)

mfreed7 replied:
> I've spec'd this as a recursive call to "handle interest change"

Clever to do it recursively. I *think* that might work? It's different from the chromium code which handles the invoker and target separately, but as far as I can tell it should result in the same behavior. So maybe ok? One thing that seems missing is the canceling of the `pending interest change handle` in various cases. E.g. [these](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=11288;drc=6f18728664c962ba99e4d3adda42a883dfb331ee). It gets set to null in the spec language, but the task is never actually cancelled that I see. I'm thinking that `reset interest state` should also cancel the task in addition to setting it back to `null`?

foolip replied:
Ah, so the way I defined these delayed and cancelable tasks is a bit weird.

The timers always keep going and eventually runs the step to queue a task, and that task always runs, but the first step is to check if it was "canceled" and do nothing. And it's considered canceled if the  delayed interest task handle isn't the same unique handle as it was set to in the outer steps (like a closure) so setting it to null (or another handle) cancels the task.

I don't love it, but it's based on setTimeout but slightly simpler in how the handle is stored.

foolip committed:
Don't reset interest state if attribute values didn't change

foolip committed:
Call lose interest from gain interest as needed

foolip committed:
Expand on interestfor summary

lukewarlow reviewed (changes requested):
Should be updated to make use of the new [Reflect] extended attribute

foolip committed:
Merge remote-tracking branch 'origin/main' into foolip/interesttarget

foolip committed:
Use [Reflect="interestfor"]

foolip committed:
Run an interest ready check after firing events

foolip committed:
Use interest ready check in one more place

foolip committed:
Gain interest on long press

foolip committed:
Call out optional context menu item

foolip committed:
Lose interest in the hide popover steps

keithamus reviewed (commented):
Some small comments for now, I have to pause reviewing now but will come back for a more thorough review later.

foolip commented on source:
```diff
@@ -88337,6 +88382,386 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
   </ol>
 
 
+  <h3 split-filename="interestfor">Interest invokers</h3>
```
Sure, how about simply "The `interestfor` attribute"?

foolip replied:
Renamed.

foolip committed:
Limit to 100 columns and drop one "invoker" instance

foolip commented on source:
```diff
@@ -88337,6 +88382,386 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <h4>The <code data-x="attr-interestfor">interestfor</code> attribute</h4>
+
+  <p>The <dfn element-attr for="html-global"><code data-x="attr-interestfor">interestfor</code></dfn>
```
Done, and looked through the other long lines found with `git diff origin/main | grep -E '^\+.{101}'`. They're the same ones that you found, so I guess you have a script?

foolip committed:
Rename the overall section

foolip committed:
Don't use the term "invoker"

foolip committed:
Lose interest as part of the close request steps

mfreed7 commented on source:
```diff
@@ -88337,6 +88401,383 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  event fired on it. If the target is a popover with a <span>popover visibility state</span> of
+  <span data-x="popover-hidden-state">hidden</span>, this will show the popover. When the user loses
+  interest (e.g., by no longer hovering or focusing the interest source or target) a <code
+  data-x="event-loseinterest">loseinterest</code> event is fired. If the target is a popover it will
```
Maybe obvious, but should you add "if the target is a popover with the visibility state of showing" before it gets hidden?

mfreed7 commented on source:
```diff
@@ -88337,6 +88401,383 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  a <code data-x="attr-popover">popover</code> representing the tooltip.</p>
+
+  <pre><code class="html">&lt;button interestfor=tooltip>
+ Click me
```
Perhaps instead of "Click me" it should say "Hover or focus me to show the tooltip" or something?

mfreed7 commented on source:
```diff
@@ -88337,6 +88401,383 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   data-x="event-contextmenu">contextmenu</code> event fired, or null if no such event was
+   fired.</p></li>
+
+   <li><p>If <var>element</var> is not an <code>a</code>, <code>area</code>, or <code>button</code>
```
This can likely be moved up a step, before looking at `event`.

foolip replied:
Because "fire any relevant events" is a bit unusual, and a workaround for the fact that there's no algorithm in https://w3c.github.io/uievents/ that can be modified, I wanted to keep it close together with the step that gets the event. This like in https://html.spec.whatwg.org/#close-request-steps, where the steps are together.

Instead of moving steps around, I went a step further and set a _showContextMenu_ as early as possible, separating the event stuff from the rest. I think it's also clearer this way, and one could add optional UA-specific steps to adjust _showContextMenu_ if there are additional rules beyond the "contextmenu" event.

Hope you like it?

mfreed7 reviewed (approved):
I added a few comments, but nothing major. Overall this looks pretty good to me! And I think it matches the Chromium implementation fairly closely. I took a few notes on areas that I think might need WPT updates, and I'll get those landed. But otherwise, LGTM!

foolip committed:
Merge remote-tracking branch 'origin/main' into foolip/interesttarget

foolip committed:
Introduce gainInterestIfReady for consistent context menu behavior

foolip committed:
Introduce active interest sources set for Esc handling

foolip committed:
Improve example

foolip committed:
Elaborate in intro

foolip committed:
Hoist a showContextMenu variable for clarity

foolip commented on source:
```diff
@@ -84260,6 +84301,25 @@ body { display:none }
    </li>
 
+   <li>
+    <p>For each element <var>element</var> in <var>document</var>'s <span>shadow-including
```
OK, I've introduced a "active interest sources set", it's very easy to maintain. Note that I traversed it in reverse order here, to me it makes the most sense to try to lose interest in the oppose order that it was gained, since we're picking a specific order.

Just a note that if we want to revert this and go with tree order, Blink has `TreeOrderedMap` (used for IDs, it seems) and also a `CustomElementUpgradeSorter` that do similar things.

foolip committed:
Define :interest-source and :interest-target for HTML

https://drafts.csswg.org/selectors/#interest-pseudos

foolip commented:
I have now pushed definitions of when `:interest-source` and `:interest-target` should match. Those pseudo-classes are now defined in https://drafts.csswg.org/selectors/#interest-source.

foolip committed:
Properly link interest-delay-* properties

foolip committed:
Merge remote-tracking branch 'origin/main' into foolip/interesttarget

foolip committed:
Check ancestor chain in "handle interest change"

foolip commented on source:
```diff
@@ -88918,6 +88920,26 @@ dictionary <dfn dictionary>InterestEventInit</dfn> : <span>EventInit</span> {
+     <li><p>If <var>current</var>'s <span>active interest source</span> is not null, then return
+     that <span>active interest source</span>.</p></li>
+
+     <li><p>Set <var>current</var> to <var>current</var>'s <span>parent element</span>.</p></li>
```
@mfreed7 should this really be the parent element? The similar popover algorithms use the flat tree, should that be used here too? If yes, can you add a test that crosses a shadow DOM boundary?

mfreed7 reviewed (approved):
This LGTM, other than the flat tree comment you made.

foolip committed:
Merge remote-tracking branch 'origin/main' into foolip/interesttarget

foolip committed:
Wrap algorithms

foolip committed:
Split task handles to allow canceling only lose interest tasks, and do that for all upstream interest sources on any "show" interest change

foolip commented:
https://github.com/whatwg/html/pull/11006/commits/fa121b61ee7bc1b96ee80e0f32b9d52f625ac895 is intended to align with https://chromium-review.googlesource.com/c/chromium/src/+/6970966. @mfreed7 can you carefully review it?

It's now closer to the Chromium implementation in that it doesn't depend on a recursive call to "handle interest change", but there are differences that I'm unsure about:

- I never cancel pending gain interest tasks, because I couldn't understand what effect it could have. All of the active interest sources already have interest (already "showing") and even if there are pending gain interest tasks, they will end up being no-ops, right?
- If I'm reading the Chromium code right, the iteration after `AllSourceInterestInvokers()` comes _after_ a check that the element is an `a`, `area`, or `button` element. In the spec I've done it before, since I can't see what else would prevent a popover from closing when moving focus or hovering from the interest source to something inside the target popover, in the simplest case just a `<div>` with some text.

mfreed7 commented:
> [fa121b6](https://github.com/whatwg/html/commit/fa121b61ee7bc1b96ee80e0f32b9d52f625ac895) is intended to align with https://chromium-review.googlesource.com/c/chromium/src/+/6970966. @mfreed7 can you carefully review it?

Thank you!

> It's now closer to the Chromium implementation in that it doesn't depend on a recursive call to "handle interest change", but there are differences that I'm unsure about:
> 
> * I never cancel pending gain interest tasks, because I couldn't understand what effect it could have. All of the active interest sources already have interest (already "showing") and even if there are pending gain interest tasks, they will end up being no-ops, right?

This seems right. I removed that code from Chromium and the tests still pass. I'll land that change, but this seems ok as-is in spec.

> * If I'm reading the Chromium code right, the iteration of `AllSourceInterestInvokers()` comes _after_ a check that the element is an `a`, `area`, or `button` element. In the spec I'm iterating "recursive active interest sources" before the same check. I can't see what else would prevent a popover from closing when moving focus or hovering from the interest source to something inside the target popover, in the simplest case just a `<div>` with some text.

Hmm, `HandleInterestForHoverOrFocus()` is the only one that calls `AllSourceInterestInvokers()`, but that can be called on any element, particularly *targets* of interest invokers, and *contents* of both invokers and targets. So it can get called on elements that aren't an `a`, `button`, etc. I just stuck a comment on the change, [here](https://github.com/whatwg/html/commit/fa121b61ee7bc1b96ee80e0f32b9d52f625ac895#r167157348), that outlines the problems that I can see. I wonder if it's worth trying to bring the spec closer to the implementation of `HandleInterestForHoverOrFocus()`?

dandclark commented on source:
```diff
@@ -91725,6 +91802,521 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>Let <var>continue</var> be the result of <span data-x="concept-event-fire">firing an
+   event</span> named <code data-x="event-interest">loseinterest</code> at <var>target</var>, using
+   <code>InterestEvent</code>, with the <code data-x="dom-Event-cancelable">cancelable</code> and
+   <code data-x="dom-Event-composed">composed</code> attributes initialized to true, and the
```
In https://github.com/whatwg/html/pull/11006/changes/966e2ef4155b9cbcca99c1584c0c315326408a64 the `interest` event was made non-composed, but it looks like `loseinterest` was missed and should also be non-composed.

mfreed7 replied:
Yep, note this comment (https://github.com/whatwg/html/pull/11006/changes#r2072165717) - I think the other one just got missed.

mfreed7 commented on source:
```diff
@@ -91725,6 +91802,521 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+   <li><p>Let <var>continue</var> be the result of <span data-x="concept-event-fire">firing an
+   event</span> named <code data-x="event-interest">loseinterest</code> at <var>target</var>, using
+   <code>InterestEvent</code>, with the <code data-x="dom-Event-cancelable">cancelable</code> and
+   <code data-x="dom-Event-composed">composed</code> attributes initialized to true, and the
```
```suggestion
   <code>InterestEvent</code>, with the <code data-x="dom-Event-cancelable">cancelable</code> attribute initialized to true, and the
```

sb3nder commented on source:
```diff
@@ -90810,6 +90883,10 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
        <li><p>Set <var>element</var>'s <span>popover close watcher</span> to null.</p></li>
       </ol>
      </li>
+
+     <li><p>If <var>element</var>'s <span>active interest source</span> is not null, then
+     <span>lose interest</span> in <var>element</var>'s <span>active interest source</span>
+     with <var>element</var>.</p></li>
```
Isn’t it really strange that interest can be lost here while the interest source is still being hovered over or focused?

sb3nder replied:
Maybe this is here because of delays? Anyway, it still feels strange to me.

If delays are the cause, I would maybe expect different logic to handle that: when de-hovering etc, if the target popover is closed, use an `interest-delay-end` of `0`.

mfreed7 replied:
I think there are a couple reasons for this bit, but the important one is light dismiss. Some interfaces (e.g. touch) allow a tap/click without a hover- or focus-loss. In those cases, you should definitely lose interest.

sb3nder replied:
What about something like this?:
Here if something is in hover or focus, don't lose interest.
And somewhere else on hover- or focus-loss, if the target popover is in the hidden state, lose interest without delay.
